src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+                field_name))
+        end
+        if field_id < min_field_id or field_id > max_field_id then
+            error(('Id %d in field %q is out of range [%d; %d]'):format(
+                field_id, field_name, min_field_id, max_field_id))
+        end
+        if field_id >= reserved_field_id_min and
+           field_id <= reserved_field_id_max then
+           error(('Id %d in field %q is in reserved ' ..
+               'id range [%d, %d]'):format(field_id, field_name,
+               reserved_field_id_min, reserved_field_id_max))
+        end
+        local field_def = {
+            type = field_type,
+            name = field_name,
+            id = field_id,
+        }
+        field_by_name[field_name] = field_def
+        field_by_id[field_id] = field_def
+    end
+    return {
+        type = 'message',
+        name = message_name,
+        field_by_name = field_by_name,
+        field_by_id = field_by_id
+    }
+end
+
+-- Create a enum object suitable to pass into
+-- the protobuf.protocol function.
+--
+-- Accepts a name of an enum and an enum definition
+-- in the following format.
+--
+-- enum_def = {
+--     <value_name> = <value_id>,
+--     <...>
+-- }
+local function enum(enum_name, enum_def)
+    local value_by_name = {}
+    local value_by_id = {}
+    for value_name, value_id in pairs(enum_def) do
+        if value_by_id[value_id] ~= nil then
+            error(('Double definition of enum field %q by %d'):format(
+                value_name, value_id))
+        end
+        value_by_name[value_name] = value_id
+        value_by_id[value_id] = value_name
+    end
+    return {
+        type = 'enum',
+        name = enum_name,
+        value_by_name = value_by_name,
+        value_by_id = value_by_id,
+    }
+end
+
+-- Create a protocol object that stores message
+-- data needed for encoding.
+--
+-- Accepts protocol definition using protobuf.message
+-- and protobuf.enum functions as in example.
+--
+-- protocol_def = {
+--     protocol.message(<message_name>, <message_def>),
+--     protocol.enum(<enumc_name>, <enum_def>),
+--     <...>
+-- }
+local function protocol(protocol_def)
+    local return_protocol = {}
+    local declarations = {}
+    for _, def in ipairs(protocol_def) do
+        local field_name = def['name']
+        local field_type = def['type']
+        if declarations[field_name] then
+            error(('Double definition of name %q'):format(field_name))
+        end
+        if field_type == 'message' then
+            for _, val in ipairs(def['field_by_id']) do
+                local val_type = val['type']
+                local standard = protobuf_types[val_type]
+                local declared = declarations[val_type]
+                if val_type == field_name then
+                    error(('Message %q has a field of %q type ' ..
+                        'Recursive definition is not allowed'):format(
+                        field_name, val_type))
```

Comment:
AFAIU, it checks direct recursion, but doesn't check indirect recursion (foo includes bar, which includes foo). If it is so, it deserves a TODO comment, I think.

Reply:
Still working on it. Probably we should leave recursive definition in connection with the specificity of Lua

Reply:
Deletes this error because of normal encoding of recursive protocol

Reply:
OK. However, AFAIK not all implementations support it. So I would like to mention it in the documentation request. Sorry for nitpicking.

Reply:
Seems solved. Please, respond to comments, when they're fixed.